Implements repository

IvanLinnik opened the pull request.

IvanLinnik committed:
add repository
add REST methods in controller
change pubDate type in dto
some refactoring

IvanLinnik committed:
rename scheduler -> quartz (just like in fb-product-fetcher project)

IvanLinnik committed:
refactoring:
add interfaces LostFilmParser.java and EpisodesConverter.java and its implementations
rename RssItem into Episode
rename methods and method parameters according to new structure

PS I hope i'm understand what do you mean about possibility of changing interface or type of incoming file.

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/dto/Episode.java:
```diff
@@ -0,0 +1,20 @@
+
+@Value
+@Builder
+public class Episode {
```
if you storing this into DB, where is an id field?

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/repository/EpisodeRepositoryImpl.java:
```diff
@@ -0,0 +1,103 @@
+    private final EpisodesConverter episodesConverter;
+
+    @Override
+    public void create(Episode episode) {
```
if this method does save or update then call it `saveOrUpdate`

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/repository/EpisodeRepositoryImpl.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @Override
+    public List<Episode> getItems() {
```
what are the items? you work with episodes

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/EpisodesConverter.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.time.LocalDateTime;
+
+public interface EpisodesConverter {
```
what is the purpose of this converter?

IvanLinnik replied:
If we will use other dto with other field names, or other format of publishing date, we just need new implementation. 
For example imdb use other "title", "id", "description". So we will create new episodeConverterImpl and new episodeRepositoryImpl with this converter.
Or i need remove this interface?

mmarusich replied:
How you will implement other converter for other DTO if all methods of this interface working with Episode? 

The proper way of creating such abstraction is to bring back RssItem which was used in LostfilmParser and then in service convert RssItem to Episode

IvanLinnik replied:
yes, can i do that? just for training.

mmarusich replied:
sure

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/LostFilmParser.java:
```diff
@@ -1,63 +1,11 @@
 
-    @Value("${lostfilm.rss.tag}")
-    private final String lostfilmRssTag;
+    List<Episode> getItemList();
```
again items

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/LostFilmService.java:
```diff
@@ -1,9 +1,25 @@
 
 import java.util.List;
 
 public interface LostFilmService {
```
this service does to many things.

1. works with lost film to get new episodes
2. Works with system episodes in DB, why? (split it to LostFilmSerive and EpisodeService)

IvanLinnik replied:
so if i do this:

public class LostFilmService implements VideoService {
    private final VideoRepository videoRepository;
    private final DataParser dataParser;
    private final EpisodesMapper episodesMapper;

    public void saveNewVideos() {
        dataParser.getNewItems()
                  .stream()
                  .map(episodesMapper::map)
                  .forEach(videoRepository::saveOrUpdate);
    }

it will get worse? for now this service work with repo, work with parser and work with mapper.

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/impl/EpisodesConverterImpl.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String getDescription(Episode episode) {
+        return episode.getDescription();
```
those are simple getters, why you need this converter?

IvanLinnik committed:
refactoring:
rename methods and method parameters according to new structure

add LostFilmRssItem.java
make EpisodesConverter.java (Item -> Episode) work and rename it to EpisodesMapper.java

add structure for future implementation:
add movies
parse it from json
add superclasses for items and videos

IvanLinnik commented:
I rename many classes for future, when will be episodes and movies.
It will be new table in DB for movies.
I try make implementation of new classes, but have problem with abstract superclass. I'm not sure if it write structure.
Also i want to make DataParser  and VideoMapper work with abstract Items and Videos.

IvanLinnik commented:
remove some crud methods, because they unneeded now and just distract. I'll return it later if its needed

IvanLinnik committed:
add logging running application

IvanLinnik committed:
remove movies and its controller, services, repo
move parsing xml and mapping (item->episode) from service in quartz scheduler
refactor episode repo

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/quartz/CheckUpdateJob.java:
```diff
@@ -0,0 +1,25 @@
+        dataParser.getNewItems()
+                  .stream()
+                  .map(episodesMapper::map)
+                  .forEach(videoService::saveEpisode);
```
👍

IvanLinnik replied:
write now i don't understand this absolutely clearly, but i will

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/impl/EpisodesMapper.java:
```diff
@@ -0,0 +1,19 @@
+import io.skai.okta.internshipstreamingserviceparserxml.dto.LostFilmRssItem;
+import org.springframework.stereotype.Service;
+
+@Service
```
@Component

mmarusich commented on src/main/resources/db/changelog/db.changelog-master.xml:
```diff
@@ -16,4 +16,30 @@
+    <changeSet author="Linnik Ivan" id="episodesAddColumns1">
+        <addColumn tableName="episodes">
+            <column name="season" type="INTEGER"/>
+            <column name="episodeNumber" type="INTEGER"/>
```
column names in DB should be lower case, in you case `episode_number`

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/impl/LostFilmService.java:
```diff
@@ -0,0 +1,60 @@
+        return videoRepository.get(link);
+    }
+
+//    @Override
```
clean it, don't keep commented code

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/repository/impl/EpisodeRepository.java:
```diff
@@ -0,0 +1,115 @@
+        if (isPresent(episode)) {
+            log.info("update");
+            update(episode);
+        } else {
```
instead of if/else statement do return 
`if(sss) {do something; return} do something else`

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/repository/impl/EpisodeRepository.java:
```diff
@@ -0,0 +1,115 @@
+    @Override
+    public void saveOrUpdate(Episode episode) {
+        if (isPresent(episode)) {
+            log.info("update");
```
this log not informative, it does not give you any useful information

IvanLinnik replied:
ooops. it was just for debugging. i'm delete it now.

IvanLinnik committed:
move mapper to separate folder
add work with new fields in repository
add new fields in mapper
update jooq autogenerated files
some refactoring

IvanLinnik committed:
cleanup commented lines
rename columns in db
refactoring according to review

IvanLinnik committed:
refactoring EpisodesMapper.java

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/mapper/EpisodesMapper.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private String extract(String from, String value) {
+        final String[] result = from.replaceAll("\\(", ")").split("\\)");
```
more proper way is to use regex for extracting parts from string

mmarusich commented on src/main/java/io/skai/okta/internshipstreamingserviceparserxml/services/mapper/EpisodesMapper.java:
```diff
@@ -0,0 +1,40 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class EpisodesMapper {
```
it's a classic converter

mmarusich reviewed (approved):
LGTM, just fix title parsing using regex, you can do it in separate PR

IvanLinnik committed:
rename EpisodesMapper.java to EpisodeConverter.java
